storage/src/key_value/agent/block.rs

```diff
@@ -0,0 +1,364 @@
+            index -= step;
+        }
+        Ok(block_locator_hashes)
+    }
+
+    pub(super) fn find_sync_blocks(
+        &mut self,
+        block_locator_hashes: Vec<Digest>,
+        block_count: usize,
+    ) -> Result<Vec<Digest>> {
+        let mut min_hash = None;
+        for hash in block_locator_hashes.iter() {
+            match self.get_block_state(hash)? {
+                BlockStatus::Committed(_) => {
+                    min_hash = Some(hash.clone());
+                    break;
+                }
+                _ => (),
+            }
+        }
+        let min_height = if min_hash.is_none() {
+            0
+        } else {
+            let min_hash = min_hash.unwrap();
+            let min_height = self.get_block_state(&min_hash)?;
+            match min_height {
+                BlockStatus::Committed(n) => n + 1,
+                _ => return Err(anyhow!("illegal block state")),
+            }
+        };
+        let mut max_height = min_height + block_count;
+        let canon = self.canon()?;
+        if canon.block_height < max_height {
+            max_height = canon.block_height;
+        }
+        let mut hashes = vec![];
+        for i in min_height..=max_height {
+            hashes.push(self.get_block_hash_guarded(i as u32)?);
+        }
+        Ok(hashes)
+    }
+
+    pub(super) fn get_transaction_location(&mut self, transaction_id: &Digest) -> Result<Option<TransactionLocation>> {
+        let location = self.inner.get(KeyValueColumn::TransactionLookup, &transaction_id[..])?;
+        match location {
+            Some(location) => Ok(Some(TransactionLocation::read(&location[..])?)),
+            None => Ok(None),
+        }
+    }
+
+    pub(super) fn get_canon_blocks(&mut self, limit: Option<u32>) -> Result<Vec<SerialBlock>> {
+        let index = self.inner.get_column(KeyValueColumn::BlockIndex)?
+            .into_iter()
+            .filter(|(key, _)| key.len() == 4) // only interested in block index -> block hash maps
+            .map(|(key, value)| (Self::read_u32(&key[..]).expect("invalid key"), value[..].into()))
+            .collect::<Vec<(u32, Digest)>>();
+        let mut blocks: HashMap<u32, SerialBlock> = HashMap::new();
+        let mut max_block_number = 0u32;
+        for (block_number, hash) in index {
+            if let Some(limit) = limit {
+                if block_number > limit {
+                    continue;
+                }
+            }
+            if block_number > max_block_number {
+                max_block_number = block_number;
+            }
+            let block = self.get_block(&hash)?;
+            blocks.insert(block_number, block);
+        }
+        let mut out = Vec::with_capacity(max_block_number as usize + 1);
+        for i in 0..=max_block_number {
+            out.push(
+                blocks
+                    .remove(&i)
+                    .ok_or_else(|| anyhow!("missing block {} for get_canon_blocks", i))?,
+            );
+        }
```

Comment:
wouldn't it be simpler (and likely faster) to collect the blocks and their indices from `index` into a `Vec` to begin with and `sort_unstable_by_key`, with `key` being the block index? index contiguity could also be checked by doing a single linear pass afterwards.